app/gui/view/graph-editor/src/builtin/visualization/native/text_visualization/text_provider.rs

```diff
@@ -42,11 +45,460 @@ pub trait TextProvider {
+        resize_and_insert(&update.row_names, &mut self.row_names);
+    }
+
+    /// Return whether the given line of the GridView should be a divider.
+    /// This assumes a divider at the start and the end of the table, and a divider between each
+    /// row. The divider is one line high.
+    pub fn is_row_divider(&self, line: usize) -> bool {
+        self.iter_row_items_per_line().nth(line) == Some(TableItem::Divider)
+    }
+
+    /// Return the width of the content column at the given index in chars.
+    fn column_width(&self, index: usize) -> usize {
+        let column_width = self.columns.get(index).copied().flatten().unwrap_or_default();
+        let heading_width = self
+            .column_names
+            .get(index)
+            .cloned()
+            .flatten()
+            .map(|x| x.chars().count())
+            .unwrap_or_default();
+        column_width.max(heading_width)
+    }
+
+    /// Return the height of the content row at the given index in lines.
+    fn row_height(&self, index: usize) -> usize {
+        let row_height = self.rows.get(index).copied().flatten().unwrap_or_default();
+        let heading_height = self
+            .column_names
+            .get(index)
+            .cloned()
+            .flatten()
+            .map(|x| x.lines().count())
+            .unwrap_or_default();
+        row_height.max(heading_height)
+    }
+
+    /// Iterate over the item in the table. Only returns the actual content items. Does not return
+    /// the dividers or the headings.
+    fn iter_content_columns(&self) -> impl Iterator<Item = TableItem> + '_ {
+        self.columns.iter().enumerate().map(|(index, _)| TableItem::Content {
+            size: self.column_width(index),
+            index,
+            offset: 0,
+        })
+    }
+
+    /// Iterate over the item in the table. Only returns the actual content items. Does not return
+    /// the dividers or the headings.
+    fn iter_content_rows(&self) -> impl Iterator<Item = TableItem> + '_ {
+        self.rows.iter().enumerate().map(|(index, _)| TableItem::Content {
+            size: self.row_height(index),
+            index,
+            offset: 0,
+        })
+    }
+
+    /// Return the number of content rows. This does noit count the row headings or the row
+    /// dividers.
+    pub fn content_row_count(&self) -> usize {
+        self.rows.len()
+    }
+
+    /// Return the number of content columns. This does not count the column headings or the column
+    /// dividers.
+    pub fn content_column_count(&self) -> usize {
+        self.columns.len()
+    }
+
+    /// Return the width of the table in characters.
+    pub fn width_in_chars(&self) -> usize {
+        let res = self.iter_column_items().map(|x| x.size()).sum();
+        res
+    }
+
+    /// Convert the given cell / chunk position from the backend data into a chunk position for the
+    /// text cache.
+    pub fn backend_table_position_to_grid_content_position(
```

Comment:
It might be a good idea to add information about time complexity - here and in next methods. We iterate over all columns on the left (if I understand the code correctly), and that might be substantial.

Reply:
Or are those _displayed_ columns only? If yes, then I miss that important information here or in the structure's docs.